### ISSUE TITLE
Support to access EditorEventManager from LanguageClient implementation

### ISSUE THREAD

**Issue by gayanper** [ballerina-platform/lsp4intellij#46]
**Description:**
When adding support for custome notifications (language server specific) we need to implement our own language client implementation where we can add JsonNotification annotations. But we also need to access to components like EditorEventManager for example to resolve the current editor for uri. So it would be nice to have a client context injected with these required components into the LanguageClient implementation. So that custom implementations can use it, even when implementing the codelenses this might be usefull to have this context.

@NipunaRanasinghe and @nixel2007 what do you think about this design idea ?

**Comment by gayanper** [comment]
Check commit https://github.com/gayanper/lsp4intellij/commit/59c21ad64b6ccc7eb7dc0655b4e780c147f0c9b3 for a initial WIP implementation of this idea.

**Comment by NipunaRanasinghe** [comment]
@gayanper Sorry for not being around here for a while. Based on my experience, to extend this language client, we need to access(extend) the following classes.
- `EditorEventManager`, 
- `DefaultRequestManager` and 
-`Languageserver` (if we need to define the remote server interface) 

Thats why I've added `LSPExtensionManager` interface so that library users can implement their own language server protocol extensions. (Please refer https://github.com/NipunaRanasinghe/lsp4intellij-plugin/blob/master/src/main/java/com/github/lsp4intellijplugin/ballerinaextension/BallerinaLSPExtensionManager.java as a reference implementation for ballerina language server). @gayanper Could you please go through it add add your suggestions on the approach?? If it is good enough, we can work on the improvements :) WDYT?

**Comment by gayanper** [comment]
@NipunaRanasinghe well i don't really understand how this extension support will solve the my issue, because i need a bunch of custom notifications on the language client. For the server it is doable since we create a proxy for the remote interface, but client has a concrete class. I don't see how we can do this without extending the language client implementation. This is what i want to implement using the lsp4intellij https://github.com/gayanper/sts4/blob/master/idea-extensions/idea-boot-java/src/org/spring/tools/boot/java/ls/StsLanguageClientImpl.scala

**Comment by gayanper** [comment]
@rasika do you have any design suggestion on this ?

**Comment by gayanper** [comment]
Any update on this guys. I would like if some one can answer my last question, so i can work on the changes in this coming long weekend 😉

**Comment by rasika** [comment]
@gayanper Based on @NipunaRanasinghe's https://github.com/ballerina-platform/lsp4intellij/issues/46#issuecomment-483495679 we can allow extending [LanguageClientImpl](https://github.com/ballerina-platform/lsp4intellij/blob/24c72567ab87686a885793df860ecc411ca4e95c/src/main/java/org/wso2/lsp4intellij/client/LanguageClientImpl.java) through the [LSPExtensionManager](https://github.com/ballerina-platform/lsp4intellij/blob/24c72567ab87686a885793df860ecc411ca4e95c/src/main/java/org/wso2/lsp4intellij/extensions/LSPExtensionManager.java). For example(LSPExtensionManager.java);
```java
LanguageClient getExtendedLanguageClient();
```

Regarding the context implementation; I referred your implementation in gayanper@59c21ad. +1 for it with the below change.

Instead of updateContext(), let's pass the context for the LanguageClientImpl through the constructor.  Reason is that ClientContext should not be NULL through-out the lifecycle of the LanguageClientImpl instance to avoid NPEs.

eg. 
```java
public LanguageClientImpl(ClientContext context) {
    this.context = context;
}
```

**Comment by gayanper** [comment]
@rasika, agree I didn't went for constructor change because i thought it will change lot of initialization related to where the client is created and stuff. But i see its minor and non risky change. If we move the LanguageClient to be taken by the extension. What should we do about the `LanguageServerDefinition.createLanguageClient` method ? Should be remove it and may be use a factory which makes the decision from where to take the language client ?

**Comment by rasika** [comment]
Yes we can remove `LanguageServerDefinition.createLanguageClient` since it just initializes the `LanguageClientImpl`. Further, we need to get default `LanguageClientImpl()` when `LSPExtensionManager || LSPExtensionManager.getExtendedLanguageClient()` is NULL.

**Comment by NipunaRanasinghe** [comment]
@rasika Currently we are allowing the user to extend only the  `DefaultRequestManager` since it does all of the LSP request handling (including the `LanguageClient` requests ). But as you suggested we can change the implementation to get a ExtendedLanguageClient through the LSPExtentionManager as well.  Shall we proceed with that approach??

@gayanper I think we can proceed with your suggestion to replace `LanguageServerDefinition.createLanguageClient` with a provider which allows to inject extended/custom language clients.

**Comment by gayanper** [comment]
@NipunaRanasinghe to make things aligned what if we rename the LanguageClientImpl as DefaultLanguageClient ?

**Comment by NipunaRanasinghe** [comment]
@gayanper +1 :)

**Comment by NipunaRanasinghe** [comment]
@gayanper As per our gitter chat, could you please check if we've lose any other commit apart from removing LanguageClientImpl? :) I can replace the class but not sure whether it'll affect your implementation.. If  not, I can proceed.

**Comment by gayanper** [comment]
@NipunaRanasinghe every thing is there, only thing that remains is the duplicate Impl class.

**Comment by NipunaRanasinghe** [comment]
Closing the issue since fixed with #52 and #62. @gayanper Thanks a lot for working on this!
